crates/rattler_index/src/lib.rs

```diff
@@ -97,117 +127,319 @@ pub fn package_record_from_conda(file: &Path) -> std::io::Result<PackageRecord>
+            subdir
+        );
+    }
+    let uploaded_packages: HashSet<String> = op
+        .list_with(&format!("{}/", subdir.as_str()))
+        .await?
+        .iter()
+        .filter_map(|entry| {
+            if entry.metadata().mode().is_file() {
+                let filename = entry.name().to_string();
+                // Check if the file is an archive package file.
+                ArchiveType::try_from(&filename).map(|_| filename)
+            } else {
+                None
+            }
         })
         .collect();
 
-    // find all subdirs
-    let mut platforms = entries
+    tracing::debug!(
+        "Found {} already uploaded packages in subdir {}.",
+        uploaded_packages.len(),
+        subdir
+    );
+
+    let packages_to_delete = registered_packages
+        .keys()
+        .cloned()
+        .collect::<HashSet<_>>()
+        .difference(&uploaded_packages)
+        .cloned()
+        .collect::<Vec<_>>();
+
+    tracing::debug!(
+        "Deleting {} packages from subdir {}.",
+        packages_to_delete.len(),
+        subdir
+    );
+
+    for filename in packages_to_delete {
+        registered_packages.remove(&filename);
+    }
+
+    let packages_to_add = uploaded_packages
+        .difference(&registered_packages.keys().cloned().collect::<HashSet<_>>())
+        .cloned()
+        .collect::<Vec<_>>();
+
+    tracing::debug!(
+        "Adding {} packages to subdir {}.",
+        packages_to_add.len(),
+        subdir
+    );
+
+    let pb = Arc::new(progress.add(ProgressBar::new(packages_to_add.len() as u64)));
+    let sty = ProgressStyle::with_template(
+        "[{elapsed_precise}] {bar:40.cyan/blue} {pos:>7}/{len:7} {msg}",
+    )
+    .unwrap()
+    .progress_chars("##-");
+    pb.set_style(sty);
+
+    let tasks = packages_to_add
         .iter()
-        .filter_map(|(p, _)| {
-            p.parent().and_then(Path::file_name).and_then(|file_name| {
-                let name = file_name.to_string_lossy().to_string();
-                if name == "src_cache" {
-                    None
-                } else {
-                    Some(name)
+        .map(|filename| {
+            tokio::spawn({
+                let op = op.clone();
+                let filename = filename.clone();
+                let pb = pb.clone();
+                let semaphore = semaphore.clone();
+                {
+                    async move {
+                        let _permit = semaphore
+                            .acquire()
+                            .await
+                            .expect("Semaphore was unexpectedly closed");
+                        pb.set_message(format!(
+                            "Indexing {} {}",
+                            subdir.as_str(),
+                            console::style(filename.clone()).dim()
+                        ));
+                        let file_path = format!("{subdir}/{filename}");
+                        let buffer = op.read(&file_path).await?;
```

Comment:
The returning `Buffer` implements `bytes::Buf` so we can use:

```rust
package_record_from_tar_bz2_reader(buffer.reader())
```

In this way, we can avoid some extra copy of data.

Reply:
thanks! implemented in #1123